File: backend.spring/src/test/scala/com/skewwhiffy/auraltester/fractions/FractionTest.scala
package com.skewwhiffy.auraltester.fractions

import org.scalatest.flatspec.AnyFlatSpec

class FractionTest extends AnyFlatSpec {
  it should "instantiate simple fraction" in {
    val expected = "1/2"

    val actual = Fraction(1, 2)

    assert(actual.toString == expected)
  }

  it should "instantiate mixed fraction" in {
    val expected = "1 2/3"

    val actual = Fraction(1, 2, 3)

    assert(actual.toString == expected)
  }

  it should "add fractions" in {
    val expected = "11/12"

    val actual = Fraction(3, 4) + Fraction(1, 6)

    assert(actual.toString == expected)
  }

  it should "subtract fractions" in {
    val expected = "1/2"

    val actual = Fraction(3, 4) - Fraction(1, 4)

    assert(actual.toString == expected)
  }

  it should "multiply fractions" in {
    val expected = "3/8"

    val actual = Fraction(1, 4) * Fraction(3, 2)

    assert(actual.toString == expected)
  }

  it should "equate integer fractions to integers" in {
    val source = Fraction(4, 2)

    //noinspection ComparingUnrelatedTypes
    assert(source == 2)
  }

  it should "equate equivalent fractions" in {
    val first = Fraction(2, 4)
    val second = Fraction(-3, -6)

    assert(first == second)
  }
}

File: backend/src/test/kotlin/com/skewwhiffy/fraction/FractionTest.kt
package com.skewwhiffy.fraction

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class FractionTest {
  @Test
  fun `instantiates simple fraction`() {
    val expected = "1/2"

    val actual = Fraction(1, 2)

    assertThat(actual.toString()).isEqualTo(expected)
  }

  @Test
  fun `instantiates mixed fraction`() {
    val expected = "1 2/3"

    val actual = Fraction(1, 2, 3)

    assertThat(actual.toString()).isEqualTo(expected)
  }

  @Test
  fun `adds fractions`() {
    val expected = "11/12"

    val actual = Fraction(3, 4) + Fraction(1, 6)

    assertThat(actual.toString()).isEqualTo(expected)
  }

  @Test
  fun `subtracts fractions`() {
    val expected = "1/2"

    val actual = Fraction(3, 4) - Fraction(1, 4)

    assertThat(actual.toString()).isEqualTo(expected)
  }

  @Test
  fun `multiplies fractions`() {
    val expected = "3/8"

    val actual = Fraction(1, 4) * Fraction(3, 2)

    assertThat(actual.toString()).isEqualTo(expected)
  }

  @Test
  fun `equates integer fractions to integers`() {
    val source = Fraction(4, 2)

    @Suppress("AssertBetweenInconvertibleTypes")
    assertThat(source).isEqualTo(2)
  }

  @Test
  fun `equates equivalent fractions`() {
    val first = Fraction(2, 4)
    val second = Fraction(-3, -6)

    assertThat(first).isEqualTo(second)
  }
}